cloudbridge/cloud/providers/aws/services.py

```diff
@@ -597,103 +409,60 @@ def _process_block_device_mappings(self, launch_config, zone=None):
+                ebs_def['DeleteOnTermination'] = device.delete_on_terminate
                 if device.size:
-                    bd_type.size = device.size
+                    ebs_def['VolumeSize'] = device.size
+                if ebs_def:
+                    bdm['Ebs'] = ebs_def
             else:  # device is ephemeral
-                bd_type.ephemeral_name = 'ephemeral%s' % ephemeral_counter
+                bdm['VirtualName'] = 'ephemeral%s' % ephemeral_counter
+            # Append the config
+            bdml.append(bdm)
 
-        return bdm
+        return bdml
 
     def create_launch_config(self):
         return AWSLaunchConfig(self.provider)
 
     def get(self, instance_id):
-        """
-        Returns an instance given its id. Returns None
-        if the object does not exist.
-        """
-        try:
-            reservation = self.provider.ec2_conn.get_all_reservations(
-                instance_ids=[instance_id])
-            return (AWSInstance(self.provider, reservation[0].instances[0])
-                    if reservation else None)
-        except EC2ResponseError as ec2e:
-            if ec2e.code == 'InvalidInstanceID.NotFound':
-                return None
-            elif ec2e.code == 'InvalidParameterValue':
-                # Occurs if id does not start with 'inst-...'
-                return None
-            raise ec2e
+        return self.iface.get(instance_id)
 
     def find(self, name, limit=None, marker=None):
-        """
-        Searches for an instance by a given list of attributes.
-
-        :rtype: ``object`` of :class:`.Instance`
-        :return: an Instance object
-        """
-        filtr = {'tag:Name': name}
-        reservations = self.provider.ec2_conn.get_all_reservations(
-            filters=filtr,
-            max_results=limit,
-            next_token=marker)
-        instances = [AWSInstance(self.provider, inst)
-                     for res in reservations
-                     for inst in res.instances]
-        return ServerPagedResultList(reservations.is_truncated,
-                                     reservations.next_token,
-                                     False, data=instances)
+        return self.iface.find(filter_name='tag:Name', filter_value=name,
+                               limit=limit, marker=marker)
 
     def list(self, limit=None, marker=None):
-        """
-        List all instances.
-        """
-        reservations = self.provider.ec2_conn.get_all_reservations(
-            max_results=limit,
-            next_token=marker)
-        instances = [AWSInstance(self.provider, inst)
-                     for res in reservations
-                     for inst in res.instances]
-        return ServerPagedResultList(reservations.is_truncated,
-                                     reservations.next_token,
-                                     False, data=instances)
+        return self.iface.list(limit=limit, marker=marker)
```

Comment:
This will include all instances (running, stopped, launching shutting down or terminated). Do we want that or should a filter for running ones only?

Reply:
Hmm... Not sure. We should probably list stopped, because we could restart it. Shutting down too probably, since that could fail perhaps? Terminated  is probably ok either way because our handling for this is not too bad at the moment - because we check for both terminated, and error statuses in wait_for etc.

Reply:
I guess it boils down to consistency across providers. terminated instances won't show up on OpenStack so if they're included here will that mean we'll force clients to filter the results to get consistent behavior? Such filtering is likely required because of instances that may be shutting down or initializing anyhow so maybe including all is the right thing to do and just adding a comment about this...

Reply:
Hmm... Interesting point.  Yeah, I guess it'll be quite difficult to achieve 100% consistency in these cases, since multiple provider states are mapped to the same CloudBridge state. Probably docs are the way to go.